### ISSUE TITLE
Add ColorfulTabs feature to new Tree Style Tab WebExtension

### ISSUE THREAD

**Issue by MurzNN** [piroor/treestyletab#1645]
Firefox have very useful [ColorfulTabs](https://addons.mozilla.org/ru/firefox/addon/colorfultabs/) extension that colorize tabs per domain. It works well with XUL version of Tree Style Tab extension, but new Tree Style Tab WebExtension implement his own tab bar, that cannot be styled via other addons.

So will be good to add in new Tree Style Tab WebExtension same build-in feature for colorize tabs in per-domain basis. What do you think about this?

**Comment by piroor** [comment]
Sorry I won't implement it as a built-in feature of TST. Instead I recommend you to create new addon which can work with TST. TST now provides some APIs to register stylesheet to the sidebar from other addons, and tabs in the sidebar have an attribute `data-current-uri`. So, if you [register something generated style rules](https://github.com/piroor/treestyletab/wiki/API-for-other-addons#register-and-unregister-your-addon-to-tst) like following, it will work like the legacy ColorfulTabs:

```css
.tab[data-current-uri^="https://www.google.com/"] { background: green; }
.tab[data-current-uri^="https://www.mozilla.com/"] { background: red; }
```

**Comment by jjaksic** [comment]
There is now this fork: https://addons.mozilla.org/en-US/firefox/addon/tree-style-tab-rainbowtabsfork/

**Comment by piroor** [comment]
@jjaksic Wow! By the way, now TST has some APIs to help implementing separate addon for the feature. For example [you can register custom CSS for the sidebar](https://github.com/piroor/treestyletab/wiki/API-for-other-addons#register-and-unregister). How about develop the feature as a separate addon?

**Comment by MurzNN** [comment]
@meph1570 can you lookup a TST API for possibility to make Colored tabs as TST plugin, instead of separate forked extension? This will be easier to support, instead of forking and monitor all new commits.

**Comment by MurzNN** [comment]
@piroor Can you provide some short code example of TST API usage for coloring single tab, based on hostname? At now I see only examples how to alter css only for specific tabs (all, hover, active, etc).

**Comment by piroor** [comment]
@meph1570 @MurzNN  OK, here is sample codes to set color of TST's tabs from ColorfulTabs:

```javascript
(async () => {
  const kTST_ID = 'treestyletab@piro.sakura.ne.jp';

  // define colors
  browser.runtime.sendMessage(kTST_ID, {
    type: 'register-self',
    style: `
      .tab.colorfultabs-group-1 {
        background-image: linear-gradient(to right, rgba(255, 255, 255, 0.7),
                                                    rgba(183, 144, 228, 0.5),
                                                    rgb(183, 144, 228)),
                          linear-gradient(to left, rgb(183, 144, 228),
                                                   rgb(183, 144, 228));
      }
      .tab.colorfultabs-group-2 {
        background-image: linear-gradient(to right, rgba(255, 255, 255, 0.7),
                                                    rgba(228, 196, 144, 0.5),
                                                    rgb(228, 196, 144)),
                          linear-gradient(to left, rgb(228, 196, 144),
                                                   rgb(228, 196, 144));
      }
    `
  });

  // set classes for tabs
  browser.runtime.sendMessage(kTST_ID, {
    type:  'add-tab-state',
    tabs:  (await browser.tabs.query({ url: '*://piro.sakura.ne.jp/*' })).map(tab => tab.id),
    state: 'colorfultabs-group-1'
  });
  browser.runtime.sendMessage(kTST_ID, {
    type:  'add-tab-state',
    tabs:  (await browser.tabs.query({ url: '*://addons.mozilla.org/*' })).map(tab => tab.id),
    state: 'colorfultabs-group-2'
  });

  // update color definitions
  browser.runtime.sendMessage(kTST_ID, {
    type: 'register-self',
    style: `
      .tab.colorfultabs-group-1 {
        background-image: linear-gradient(to right, rgba(255, 255, 255, 0.7),
                                                    rgba(183, 144, 228, 0.5),
                                                    rgb(183, 144, 228)),
                          linear-gradient(to left, rgb(183, 144, 228),
                                                   rgb(183, 144, 228));
      }
      .tab.colorfultabs-group-2 {
        background-image: linear-gradient(to right, rgba(255, 255, 255, 0.7),
                                                    rgba(228, 196, 144, 0.5),
                                                    rgb(228, 196, 144)),
                          linear-gradient(to left, rgb(228, 196, 144),
                                                   rgb(228, 196, 144));
      }
      .tab.colorfultabs-group-3 {
        background-image: linear-gradient(to right, rgba(255, 255, 255, 0.7),
                                                    rgba(144, 228, 185, 0.5),
                                                    rgb(144, 228, 185)),
                          linear-gradient(to left, rgb(144, 228, 185),
                                                   rgb(144, 228, 185));
      }
    `
  });
  browser.runtime.sendMessage(kTST_ID, {
    type:  'add-tab-state',
    tabs:  (await browser.tabs.query({ url: '*://www.google.com/*' })).map(tab => tab.id),
    state: 'colorfultabs-group-3'
  });
})();
```

 * You can register your addon multiple times (without unregistering), for updating style sheet.
 * You can give custom class for each tab. You can apply custom color for each tab based on the custom class.
 * You'll need to listen `browser.tabs.onCreated` and `browser.tabs.onUpdated` to apply colors for newly opened tabs.

But there is a problems on TST:

 * TST doesn't notify "sidebar is opened" message to other addons. ColorfulTabs need to apply stylesheet after TST's sidebar is initialized, so I should add new notification message to TST for this purpose.

**Comment by piroor** [comment]
I've implemented new notification message types `sidebar-show` and `sidebar-hide`. Se also: https://github.com/piroor/treestyletab/wiki/API-for-other-addons#when-the-tst-sidebar-is-shownhidden
They messages will be available on the next release of TST.

**Comment by infinity0** [comment]
For the time being I've rebased @meph1570's fork here https://github.com/infinity0/treestyletab/ if people want to use it, but would prefer to see it as a separate extension.

**Comment by MurzNN** [comment]
@infinity0 Can you describe what is difference between your and @meph1570's fork?

**Comment by infinity0** [comment]
It's based off the latest TST and I fixed a few bugs, otherwise there's no changes.

**Comment by MurzNN** [comment]
So we can hope, that you will periodically update codebase to latest TST in your fork? Great, thanks!

**Comment by infinity0** [comment]
@MurzNN part of the reason I did that rebase was to make it easy for someone to see just exactly what that extension actually does, in terms of a single commit. That should lower the barrier for spinning this off as a separate extension.

I don't want to upload this to AMO as that would send off the wrong signal, users might start demanding me keep the patch up-to-date. I might do that, I might not, I don't want to get anyone's hopes up. The best solution is for someone else to maintain a separate extension as suggested above.

**Comment by MurzNN** [comment]
I start creating separate extension using TST API for colorizing tabs here https://github.com/MurzNN/tst-colored-tabs - this is only alpha tests at now, no working features.

**Comment by infinity0** [comment]
It's not supposed to work for you.

The Debian package works fine.

**Comment by mausalimi** [comment]
ah ok, it doesn't say it works only for *some* people.

**Comment by MurzNN** [comment]
I pushed first commit to https://github.com/MurzNN/tst-colored-tabs where tab colorize function seems works normally.

**Comment by MurzNN** [comment]
I've released 0.2 version with bugfix & improvements, let's close this issue.
